File: Competitive Programming 3/Section 1.4/Game (Chess)/10196 - Check The Check/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

char grid[12][12];
int movs[][8][2]={
    {{1,-1},{1,1}},//black pawn (1)
    {{-1,-1},{-1,1}},//white pawn (1)
    {{2,1},{-2,1},{2,-1},{-2,-1},{1,2},{-1,2},{1,-2},{-1,-2}}, // knight (1)
    {{1,1},{-1,-1},{-1,1},{1,-1}},//bishop (1..n)
    {{1,0},{-1,0},{0,1},{0,-1}}, // rook (1..n)
    {{1,1},{-1,-1},{-1,1},{1,-1},{1,0},{-1,0},{0,1},{0,-1}}//queen or king (special case [1..n and 1])
};
struct config{
    int movIndex;
    bool isN;
    config(int movIndex,bool isN){
        this->movIndex=movIndex;
        this->isN=isN;
    }
    config(){}
};
map<char,config> getMovRow = {
    make_pair('p',config(0,false)),
    make_pair('P',config(1,false)),
    make_pair('n',config(2,false)),
    make_pair('b',config(3,true)),
    make_pair('r',config(4,true)),
    make_pair('q',config(5,true)),
    make_pair('k',config(5,false))
};
bool attacks(int i,int j,config conf,char objective){
    int sz = sizeof(movs[conf.movIndex])/sizeof(int**);
    for(int mov = 0 ; mov < sz; mov++){
        int iN=i,jN=j;
        for(iN+=movs[conf.movIndex][mov][0],jN+=movs[conf.movIndex][mov][1];grid[iN][jN]=='.';iN+=movs[conf.movIndex][mov][0],jN+=movs[conf.movIndex][mov][1]){
            if(!conf.isN)
                break;
        }
        if(grid[iN][jN]==objective)
            return true;
    }
    return false;
}

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    for(int cs=1;1;cs++){
        bool differentFromPoint = false;
        for(int i = 2 ; i <= 9 ; i ++){
            for(int j = 2 ; j <= 9 ; j ++){
                cin >> grid[i][j];
                if(grid[i][j]!='.')
                    differentFromPoint=true;
            }
        }
        if(!differentFromPoint)
            return 0;
        string ans = "no king is in check.";
        for(int i = 2 ; i <= 9; i++){
            for(int j = 2 ; j <= 9 ; j ++){
                if(grid[i][j]!='.'){
                    char objective = islower(grid[i][j])?'K':'k';
                    char piece = grid[i][j]!='p' && grid[i][j]!='P' ? tolower(grid[i][j]) :grid[i][j];
                    if(attacks(i,j,getMovRow[piece],objective)){
                        ans = string(objective=='K'?"white":"black")+" king is in check.";
                        goto after;
                    }
                }
            }
        }
        after:
        cout << "Game #" << cs << ": " << ans << "\n";
    }
    return 0;
}

File: Competitive Programming 3/Section 1.4/Game (Chess)/10196 - Check The Check/Python/main.py
movs=[
    [[1,-1],[1,1]],#black pawn (1)
    [[-1,-1],[-1,1]],#white pawn (1)
    [[2,1],[-2,1],[2,-1],[-2,-1],[1,2],[-1,2],[1,-2],[-1,-2]],#knight (1)
    [[1,1],[-1,-1],[-1,1],[1,-1]],#bishop (1..n)
    [[1,0],[-1,0],[0,1],[0,-1]], # rook (1..n)
    [[1,1],[-1,-1],[-1,1],[1,-1],[1,0],[-1,0],[0,1],[0,-1]]#queen or king (special case [1..n and 1])
]
class Config:
    def __init__(self,movIndex,isN):
        self.movIndex=movIndex
        self.isN=isN
getMovRow = {
    'p':Config(0,False),
    'P':Config(1,False),
    'n':Config(2,False),
    'b':Config(3,True),
    'r':Config(4,True),
    'q':Config(5,True),
    'k':Config(5,False)
}
grid=[['\0' for i in range(12)] for j in range(12)]
def attacks(i,j,conf,objective):
    for mov in movs[conf.movIndex]:
        iN,jN = i + mov[0],j+mov[1]
        while grid[iN][jN] == '.':
            if not conf.isN:
                break
            iN+= mov[0]
            jN+= mov[1]
        if grid[iN][jN] == objective:
            return True 
    return False
def solve():
    for i in range(2,10):
        for j in range(2,10):
            if grid[i][j]!='.':
                objective = 'K' if str.islower(grid[i][j]) else 'k'
                piece = str.lower(grid[i][j]) if grid[i][j]!='p' and grid[i][j] != 'P' else grid[i][j]
                if attacks(i,j,getMovRow[piece],objective):
                    return ("white" if objective=='K' else "black") + " king is in check."
    return "no king is in check."
def main():
    cs=1
    while True:
        differentFromPoint = False
        for i in range(2,10):
            line = input()
            for j in range(2,10):
                grid[i][j] = line[j-2]
                if grid[i][j]!='.':
                    differentFromPoint=True
        if not differentFromPoint:
            break
        print("Game #"+str(cs)+": "+solve())
        input()              
        cs+=1
main()